### ISSUE TITLE
Plesk does not support our Dockerfile syntax

### ISSUE THREAD

**Issue by spannerman79** [owncast/owncast#1447]
Hello.

Background:

I'm trying to get Owncast running via Docker within Plesk.  It isn't as simple as docker-compose or just simply running docker normally .

-----

In my testing to see if indeed it was something that I screwed up on I tested portainer as it uses three ports ( see https://github.com/portainer/portainer/blob/develop/build/linux/Dockerfile#L8..L10 ) and **that** does work.

I followed the same sort of setup to get Owncast going, but it seems your `EXPOSE` combines two ports in the same line.

I've also done some other digging and experimenting to see if this is just a one off on Plesk being a pain in the you know where but it isn't.

It seems your `EXPOSE` of combining two ports on the same line causes issues - see https://github.com/owncast/owncast/blob/develop/Dockerfile#L3

Sidenote: I actually have Plesk running on a Vultr VPS as I saw on #1446

**Comment by jeyemwey** [comment]
I have seen `EXPOSE` used both ways ([multiple inline](https://stackoverflow.com/a/32806333) and [one after another](https://www.baeldung.com/ops/docker-expose-more-than-one-port#1-declaration-in-dockerfile)), so it would be fine for me if we changed it to support Plesk while keeping everything else working.

> I've also done some other digging and experimenting to see if this is just a one off on Plesk but it isn't.

Have you found any docs / forum posts that suggest that Plesk cannot read multiple ports in one `EXPOSE` statement?

I will create an PR with the change soon, so we can try it out [with the nightly builds](https://github.com/owncast/owncast/blob/develop/.github/workflows/docker-nightly.yaml).

**Comment by spannerman79** [comment]
I haven't directly found any docs per say, but I have raised a support issue with Plesk to see what they have to say directly. I don't have a support plan with Plesk so ETA is basically _whenever someone gets to it_ 🤣 

I can only go off what I have done in testing/experimenting 🤷 

One such other user that does not have any ports shown would be in the comment section of one of their [Support Docs](https://support.plesk.com/hc/en-us/articles/115003142213-Unable-to-add-Docker-Proxy-rules-in-Plesk-no-container-is-displayed)

> I will create an PR with the change soon, so we can try it out with the nightly builds.

I'll be happy to test.  Literally no rush as its just turned midnight (12am) for me (UTC+10) so it will be a couple of hours before I can actually start to test.

To be fair I wasn't expecting a reply within 30 minutes.

**Comment by jeyemwey** [comment]
> I'll be happy to test.

Sure, just build the container with the changed Dockerfile or grab the container from the GitHub container registry.

> Literally no rush as its just turned midnight (12am) for me (UTC+10) so it will be a couple of hours before I can actually start to test. To be fair I wasn't expecting a reply within 30 minutes.

We have some in the project that are UTC+2 (like me), but most are on the US/West Coast.

**Comment by spannerman79** [comment]
> Sure, just build the container with the changed Dockerfile

If i was adept with Docker (for which I am **not**) I would have prob already done that already and provided actual reproduction steps to show that Plesk is being naughty when it comes to `EXPOSE` lol

> or grab the container from the GitHub container registry.

Never used it before, nor do I have any idea where to start with that.  Github docs is about as clear as mud to me on this matter.  A litle guidance would be appreciated. I usually start reading from https://github.com/owncast/owncast#building-from-source for that sort of info but alas its not there nor located at https://owncast.online/docs/troubleshooting/  

ie `docker pull ...` `...` being whatever else is required to pull the image _with the updated Dockerfile_.  The rest I _should_ be able to decyper 😄

**Comment by gabek** [comment]
~if you look at our Quickstart documentation on https://owncast.online it should point you in the right direction.  In this case this should have what you're looking for:~

~https://owncast.online/quickstart/container/~

Whoops I think you're talking about something completely different. Ignore my comments.

**Comment by spannerman79** [comment]
Still I would like to know the correct docker pull in order to test this fix as I kindly requested in a [previous comment ](https://github.com/owncast/owncast/issues/1447#issuecomment-933041561)

**Comment by mbround18** [comment]
For anyone else who would like to pull from ghcr for the image its:

```sh
docker pull ghcr.io/owncast/owncast:nightly
```

**Comment by gabek** [comment]
I haven't had a chance to test the syntax change in the linked PR yet, I'll check it out tomorrow.

**Comment by jeyemwey** [comment]
> For anyone else who would like to pull from ghcr for the image its: `docker pull ghcr.io/owncast/owncast:nightly`

This will only work, once the PR 1448 is merged and the nightly build is built.

**Comment by mbround18** [comment]
When i posted my comment ghcr has an image available for pulling 0.o 
https://github.com/owncast/owncast/pkgs/container/owncast

unsure the contents of #1448 (please use a hashtag plus the number to link it ;)  )

UPDATE: The `EXPOSE` in a docker file doesn't really matter, its more of an aesthetic than anything when building docker images.  


<blockquote><img src="https://github.githubassets.com/images/modules/open_graph/github-logo.png" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com">Build software better, together</a></strong></div><div>GitHub is where people build software. More than 65 million people use GitHub to discover, fork, and contribute to over 200 million projects.</div></blockquote>

**Comment by gabek** [comment]
Didn't @spannerman79 say that it matters to Plesk?

**Comment by spannerman79** [comment]
> Didn't @spannerman79 say that it matters to Plesk?

Well that is what I would like to test to make sure for a fact @gabek 

I've tried other docker images but those have their `EXPOSE` on each seperate line for each port.  Owncast is the only docker image - and what I actually really want to get going - that has issue exposing ports for Plesk reverse proxy rules.

See [this](https://support.plesk.com/hc/en-us/articles/115003142213-Unable-to-add-Docker-Proxy-rules-in-Plesk-no-container-is-displayed) support doc (and the comment made by Alexey Lapshin in that same support doc link).  The same link I gave in https://github.com/owncast/owncast/issues/1447#issuecomment-933041561

**Comment by gabek** [comment]
It looks like that support doc only says that ports have to be exposed, and in our case they are. I don't see anything mentioned about not supporting particular flavors of exposing.

If somebody can verify that this indeed fixes it for Plesk I'm happy to merge in the change, but putting in changes for things that already work in hopes it might fix something else isn't the right way to go about things.

**Comment by spannerman79** [comment]
> It looks like that support doc only says that ports have to be exposed, and in our case they are. I don't see anything mentioned about not supporting particular flavors of exposing.

**If** they were I wouldn't have created this issue in the first place....



> If somebody can verify that this indeed fixes it for Plesk I'm happy to merge in the change

https://github.com/owncast/owncast/issues/1447#issuecomment-933041561

**Comment by gabek** [comment]
Unfortunately somebody who uses Plesk is going to have to test this, so you're it! You can build a Docker image locally and test your changes.

```
docker build --build-arg NAME=docker --build-arg VERSION=mytest -t my-plesk-owncast-test .
```

```
docker run my-plesk-owncast-test
```

**Comment by spannerman79** [comment]
An update;

What I did.

1. `git clone https://github.com/owncast/owncast`
2. Edited `Dockerfile` so that each port had its own `EXPOSE` line - the same thats in https://github.com/owncast/owncast/pull/1448/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557
3. `docker build --build-arg NAME=docker --build-arg VERSION=mytest -t my-plesk-owncast-test .`

Plesk is a little strange if you wish to use their Docker proxy rules (which is the whole point of this issue creation, can skip over https://owncast.online/docs/sslproxies/ fully as Plesk has 99% of that covered) so you need to start the docker image with their extension.

I filled in the correct details required which isn't much, container volume mapping, deselect automatic port mapping and start the container.  Stopped the container and went to configure port mapping as the ports don't show up in the extension until the container is ran at least once and... nope 😞 

Volume mapping is correct (on host);
```
# ls
hls  logo.svg  logs  owncast.db  owncast.db-shm  owncast.db-wal
```

I am at a literal loss now.  Plesk Support still hasn't gotten back to me so I created a thread over on their forum (still awating approval at this time, once it is https://talk.plesk.com/threads/installing-owncast-via-docker-ports-not-exposed.362455/ ) so perhaps another Plesk user might be able to shead some more light on this.

**Comment by spannerman79** [comment]
Oh my god...

I got it working!

All I did was move https://github.com/owncast/owncast/blob/develop/Dockerfile#L3 to the literal bottom of the Dockerfile.  No other changes.

I didn't split each port into its own EXPOSE.

For _whatever_ reason Plesk Docker Extension was reading the EXPOSE as part of the build script and ignoring it;

Take a look
![plesk_docker](https://user-images.githubusercontent.com/7542384/136677392-254842e1-bde7-4a4c-be12-f5e72c870d1b.PNG)


See https://github.com/owncast/owncast/pull/1463 for the exact change done.

Edit:  tried half a dozen other locations of where I put `EXPOSE` - both single and double lined.  After waiting for the slow compile times of the compile only using 1 vCPU (and on VPS with a total of 4 vCPU) I was just about to throw the towel in when I decided for the last time I'll just shove it at the end of the `Dockerfile`

<blockquote><img src="https://repository-images.githubusercontent.com/266445736/4f82fb80-a5b1-11ea-8f1c-69d316f861b5" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/owncast/owncast">owncast/Dockerfile at develop · owncast/owncast</a></strong></div><div>Take control over your live stream video by running it yourself.  Streaming + chat out of the box. - owncast/Dockerfile at develop · owncast/owncast</div></blockquote>
